tests/does-it-work/src/main.rs

```diff
@@ -38,86 +38,218 @@ fn run_test(rx: std::sync::mpsc::Receiver<()>) {
+                .arg("dtrace")
+                .arg("-l")
+                .arg("-v")
+                .arg("-n")
+                .arg("does__it*:::")
+                .stdin(Stdio::piped())
+                .stdout(Stdio::piped())
+                .spawn()
+                .expect("Could not start DTrace");
+            let output = dtrace
+                .wait_with_output()
+                .expect("Failed to read DTrace stdout");
+
+            // Kill the test thread
+            let _ = send.send(());
+
+            // Collect the actual output
+            let output = String::from_utf8_lossy(&output.stdout);
+            println!("{}", output);
+
+            // Check the line giving the full description of the probe
+            let mut lines = output.lines().skip_while(|line| !line.contains("does__it"));
+            let line = lines
+                .next()
+                .expect("Expected a line containing the provider name");
+            let mut parts = line.split_whitespace();
+            let _ = parts.next().expect("Expected an ID");
+
+            let provider = parts.next().expect("Expected a provider name");
+            assert!(
+                provider.starts_with("does__it"),
+                "Provider name appears incorrect: {}",
+                provider
+            );
+
+            let module = parts.next().expect("Expected a module name");
+            assert!(
+                module.starts_with("does_it_work"),
+                "Module name appears incorrect: {}",
+                module
+            );
+
+            let mangled_function = parts.next().expect("Expected a mangled function name");
+            assert!(
+                mangled_function.contains("does_it_work8run_test"),
+                "Mangled function name appears incorrect: {}",
+                mangled_function
+            );
+
+            // Verify the argument types
+            let mut lines = lines.skip_while(|line| !line.contains("args[0]"));
+            let first = lines
+                .next()
+                .expect("Expected a line with the argument description")
+                .trim();
+            assert_eq!(
+                first, "args[0]: uint8_t",
+                "Argument is incorrect: {}",
+                first
+            );
+            let second = lines
+                .next()
+                .expect("Expected a line with the argument description")
+                .trim();
+            assert_eq!(
+                second, "args[1]: char *",
+                "Argument is incorrect: {}",
+                second
+            );
+
+            thr.join().expect("Failed to join test runner thread");
+        }
+    }
+
+    #[cfg(target_os = "linux")]
+    mod stap {
+        use super::run_test;
+        use std::process::Stdio;
+        use std::sync::mpsc::channel;
+        use std::thread;
+
+        #[test]
+        fn test_does_it_work_stapsdt() {
```

Comment:
this is pretty different than the DTrace version; can you explain?

Reply:
Yeah, I originally wrote this using just readelf; I'd like to say I didn't want to dictate a DTrace alternative on Linux CI but honestly I just hadn't used them and thus went with the tool I knew.

I'll rewrite this to use bpftrace to make it match better.

Reply:
Okay, turns out I might have had more of a thought here than I remembered: bpftrace can output USDTs but will not output their arguments even with the verbose flag. Thus, I left the readelf usage and added a note about this into the test. I also renamed the test to be consistent and simplified its implementation a little to not undeservedly split up statically knowable parts (like checking "Provider:" and "does__it" strings separate instead of just checking for "Provider: does__it").